Expose opening directories just for traversal (O_PATH)

gereeter opened the pull request:
Closes #3741. Further bikeshedding is welcome.

daurnimator commented on lib/std/fs.zig:
```diff
@@ -774,25 +774,68 @@ pub const Dir = struct {
+    }
 
-        const flags = os.O_RDONLY | os.O_DIRECTORY | os.O_CLOEXEC;
+    fn openDirFlagsC(self: Dir, sub_path_c: [*]const u8, flags: u32) OpenError!Dir {
```
I'd either `or` `O_DIRECTORY` into flags here; or `assert` that it is set.

daurnimator commented:
Needs rebasing/updating now that sentinel types exist.

gereeter commented:
Rebased and moved `O_DIRECTORY`. I initially kept that flag out of `openDirFlagsC` because it was purely for preventing duplication of code and it felt weird to partially but not completely deduplicate the flags, but you are right that it fits the meaning of the function better.

gereeter commented:
I'm pretty sure that the test failure is the deadlock mentioned in #3751. I'd have liked to have a proper Windows test run, though.

andrewrk commented:
The code looks good. I'll run it locally on Windows when merging.

What do you think about a slight modification (that changes docs only), making `std.Dir.openDir` defined to open for *both* traversal and listing, openDirList for only listing, and openDirTraverse for only traversal?

gereeter commented:
> I'll run it locally on Windows when merging.

Great! (It did succeed on CI before the rebase, so I do full expect it to pass. But better safe than sorry, especially with confusingly documented semantics.)

> What do you think about a slight modification (that changes docs only), making `std.Dir.openDir` defined to open for *both* traversal and listing, `openDirList` for only listing, and `openDirTraverse` for only traversal?

(Technically, this affects more than docs since it would mean taking traversal away from `openDirList` if I understand correctly.) I considered something like this, but:
- Keeping `openDir` around seems like giving the user a suboptimal (too permissive and possibly unnecessarily expensive) default instead of forcing them to consider what the correct choice is, which seems against Zig philosophy.
- At that point you have 3 functions each with 3 variants, and I'd think it would make more sense to curb that explosion by passing a set of flags like the underlying OS APIs do (which I wouldn't be opposed to, but see the next point).
- You basically never want to list the contents of a directory without also traversing it. The only use-cases I can think of are reimplementing `ls` and glob expansion, and both of those support recursive enumeration that would require traversal anyway. (And most cases where people do in practice list separately from traversal they are introducing ToCToU bugs since they open the directory twice and it may have changed.)

However, if you want me to move to a flag-based API or something like that, I can.

andrewrk commented:
Alright, I'm convinced to merge it as is. Thanks for taking the time to explain this reasoning.
